File: sample/flutter_base/packges/base_preferences/lib/src/BasePreferences.dart
import 'dart:async';
import 'package:shared_preferences/shared_preferences.dart';

abstract class PreferencesProtocol {
  Future<bool> setInt(String key, int value);
  Future<int> getInt(String key);

  Future<bool> setDouble(String key, double value);
  Future<double> getDouble(String key);

  Future<bool> setBool(String key, bool value);
  Future<bool> getBool(String key);

  Future<bool> setString(String key, String value);
  Future<String> getString(String key);
}

class BasePreferences implements PreferencesProtocol {
  static late final BasePreferences sInstance = BasePreferences();
  static BasePreferences getInstance() => sInstance;

  Future<bool> setInt(String key, int value) async {
    final prefs = await SharedPreferences.getInstance();
    return prefs.setInt(key, value);
  }

  Future<int> getInt(String key) async {
    final prefs = await SharedPreferences.getInstance();
    return prefs.getInt(key) ?? 0;
  }

  Future<bool> setDouble(String key, double value) async {
    final prefs = await SharedPreferences.getInstance();
    return prefs.setDouble(key, value);
  }

  Future<double> getDouble(String key) async {
    final prefs = await SharedPreferences.getInstance();
    return prefs.getDouble(key) ?? 0;
  }

  Future<bool> setBool(String key, bool value) async {
    final prefs = await SharedPreferences.getInstance();
    return prefs.setBool(key, value);
  }

  Future<bool> getBool(String key) async {
    final prefs = await SharedPreferences.getInstance();
    return prefs.getBool(key) ?? false;
  }

  Future<bool> setString(String key, String value) async {
    final prefs = await SharedPreferences.getInstance();
    return prefs.setString(key, value);
  }

  Future<String> getString(String key) async {
    final prefs = await SharedPreferences.getInstance();
    return prefs.getString(key) ?? '';
  }
}

File: output/src/services/BasePreferences.ts
import AsyncStorage from '@react-native-async-storage/async-storage';

export interface PreferencesProtocol {
  setInt(key: string, value: number): Promise<boolean>;
  getInt(key: string): Promise<number>;
  setDouble(key: string, value: number): Promise<boolean>;
  getDouble(key: string): Promise<number>;
  setBool(key: string, value: boolean): Promise<boolean>;
  getBool(key: string): Promise<boolean>;
  setString(key: string, value: string): Promise<boolean>;
  getString(key: string): Promise<string>;
}

class BasePreferences implements PreferencesProtocol {
  private static readonly sInstance: BasePreferences = new BasePreferences();

  static getInstance(): BasePreferences {
    return BasePreferences.sInstance;
  }

  async setInt(key: string, value: number): Promise<boolean> {
    await AsyncStorage.setItem(key, value.toString());
    return true;
  }

  async getInt(key: string): Promise<number> {
    const val = await AsyncStorage.getItem(key);
    return val !== null ? parseInt(val, 10) : 0;
  }

  async setDouble(key: string, value: number): Promise<boolean> {
    await AsyncStorage.setItem(key, value.toString());
    return true;
  }

  async getDouble(key: string): Promise<number> {
    const val = await AsyncStorage.getItem(key);
    return val !== null ? parseFloat(val) : 0;
  }

  async setBool(key: string, value: boolean): Promise<boolean> {
    await AsyncStorage.setItem(key, value.toString());
    return true;
  }

  async getBool(key: string): Promise<boolean> {
    const val = await AsyncStorage.getItem(key);
    return val === 'true';
  }

  async setString(key: string, value: string): Promise<boolean> {
    await AsyncStorage.setItem(key, value);
    return true;
  }

  async getString(key: string): Promise<string> {
    const val = await AsyncStorage.getItem(key);
    return val !== null ? val : '';
  }
}

export default BasePreferences;